Detect unknown licenses #1675

akugarg opened the pull request:
Signed-off-by: akugarg <akanksha.garg2k@gmail.com>
This PR introduces a new and effective way for detection of unknown licenses by making use of n-grams.
Refer #1675 for more details.

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

akugarg committed:
Improve licenses detection accuracy of unknowns using ngrams

Signed-off-by: akugarg <akanksha.garg2k@gmail.com>

pombredanne commented on src/licensedcode/index.py:
```diff
@@ -19,14 +19,15 @@
 from intbitset import intbitset
 
-from licensedcode import SMALL_RULE
+from licensedcode import SMALL_RULE, match_unknown
```
On import per line please

pombredanne commented on src/licensedcode/match_unknown.py:
```diff
@@ -0,0 +1,77 @@
+MATCH_UNKNOWN = '6-unknown'
+
+
+def add_ngrams(automaton, tids, rule_length, unknown_ngram_length=7):
```
What is `rule_length` doing here?

akugarg replied:
We need to check if our rule_length is less than our n-gram length right ?

pombredanne commented on src/licensedcode/match_unknown.py:
```diff
@@ -0,0 +1,77 @@
+    for ngram in rule_ngrams:
+        ngram = tuple(ngram)
+        automaton.add_word(ngram, ngram)
+
```
It does not hurt to return the updated `automaton`

pombredanne commented on src/licensedcode/match_unknown.py:
```diff
@@ -0,0 +1,77 @@
+        automaton.add_word(ngram, ngram)
+
+
+def unknown_match(idx, query_run, automaton, unknown_ngram_length=7, **kwargs):
```
may be call this `match_unknowns` ?

pombredanne reviewed (changes requested):
Thanks... See some nitpickings for your review.

akugarg committed:
Improve licenses detection accuracy of unknowns using ngrams

Signed-off-by: akugarg <akanksha.garg2k@gmail.com>

pombredanne commented on src/licensedcode/index.py:
```diff
@@ -20,17 +20,20 @@
 from intbitset import intbitset
 
 from licensedcode import SMALL_RULE
+from licensedcode import match_unknown
```
This is redundant with the import below.
```suggestion
```

pombredanne commented on src/licensedcode/index.py:
```diff
@@ -874,6 +885,25 @@ def match(
+            merge=True,
+        )
+
+        original_qspan = Span(0, len(qry.tokens)-1)
```
What is `len(qry.tokens) - 1` ? coulf you assign it to a variable so that it has a name and is easier to understand?

pombredanne replied:
Actually this is clear enough

pombredanne commented on src/licensedcode/index.py:
```diff
@@ -874,6 +885,25 @@ def match(
+        )
+
+        original_qspan = Span(0, len(qry.tokens)-1)
+        matched_qspans = [m.qspan for m in matches]
```
A small comment explaining what this code blocks does would be helpful!

pombredanne reviewed (changes requested):
Thanks! there a few tiny nit pickings for your consideration and then we can merge!

AyanSinhaMahapatra commented:
This should add some tests on licenses that ideally scancode would never add (some proprietary/sdk licenses), to check the accuracy of unknown license detection. 

Linking some of them here:

- https://github.com/AppsFlyerSDK/AppsFlyerFramework/blob/master/LICENSE

pombredanne committed:
Merge remote-tracking branch 'upstream/develop' into improve_license_detection

pombredanne committed:
Use UNKNOWN_NGRAM_LENGTH constant for ngrams

Rather than using a function argument

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Remove duplicated import

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Format code

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Merge latest develop branch

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add and update licenses and detection rules

This is a batch of misc. license detection rules and a new license.
These have been mostly found thanks to the upcoming unknown license
detection.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Add unknown_licenses argument to get_licenses()

This makes it available in the CLI

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Fix typos in comments and docstring

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Use format for matched license text highlight

This is cleaner and more composable than old-style interpolation

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Enable unknown license match filtering

Add split_weak_matches() function to pre-filter weak unknown matches.
Make unknown matches eligible for filter_spurious_matches() and lower
minimum density to 0.6=5 for longer matches. Move the call for
filter_spurious_matches() earlier in the refine pipeline.

Add new filter_invalid_contained_unknown_matches() function to discard
unknown matches found inside the matched queryregion of larger regular
matches

Extract get_full_qspan_matched_text() function from
get_full_matched_text() for improve reusability. This is designed to be
called when crafting new rules absed on a match (which is what is done
with unknown matches).

Use format for matched license text highlight
This is cleaner and more composable than old-style interpolation.

Improve debug tracing of matched texts.

Apply other minor refactoring and doc impropvements

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve UnknownRule

Create unique rule id based on a checksum of the rule content

Also improve key phrases parsing for dnagling {{ {{ braces.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Make match unknown work

Create proper synthethic Rule and LicenseMatch on match and return
a match or None.

Include unknown licenses matching as an option to Index.match

Add tests

Use shorter ngrams of length 6 rather than 7 for better sensitivity
This is balanced by the addition of filters:
- Filter weak unknown matches at match time
- Filter out several weak unknown ngrams at index time

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne committed:
Improve license rule generation error reporting.

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
@akugarg @AyanSinhaMahapatra your review is welcomed!
At this stage the unknown detection works quite nicely IMHO and I am planning to merge this ASAP

pombredanne commented:
Note that you can ignore most of the rules additions, they should have been in another PR, but I could not resist adding them... I forced an unknown licenses detection on our whole test suite and this helped spots several issues that I fixed on the fly. :P

pombredanne committed:
Add new is_generic license flag #1675

Note that this is NOT YET returned in the API and outputs

Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

pombredanne commented:
I am merging now. Your feedback is still mucho welcomed on the refinements,
